rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,517 @@
+# Meta
+
+* RFC Name: Application Service Level Telemetry
+* RFC ID: 84
+* Start-Date: 2024-01-04
+* Owners: Sergey Avseyev
+* Contributors: Matt Ingenthron, Sergey Avseyev
+* Current Status: DRAFT
+
+# Summary
+
+While Capella is well instrumented for what runs in the VPC for the user, we do
+not today have much detail on how a user's deployment is operating.  This
+proposal improves that by adding simple but powerful telemetry tracking from
+the user's environment to Capella.
+
+# Motivation
+
+Couchbase Capella uplevels the user experience over a traditional Customer
+Managed environment in that *As a Service* not only basic health but even
+optimization and proactive checking for problems with the service level.  A
+challenge here is that some of the information about service health is not
+immediately available from within the Capella Data Plane; it is in the user's
+application deployment in the cloud.  That said, some users have not defined
+service levels and are not measuring them, but rather just respond when they
+receive complaints or perceive that service levels are outside normal operating
+parameters.
+
+# General Design
+
+This feature likely will not, in many cases, point directly to the cause of an
+issue.  In most cases, it will help to ensure availability of the metrics,
+particularly when other the application does to retain logs for a long enough
+period.
+
+To provide the most value at the lowest cost, two levels of metrics are
+proposed:
+
+* Capella Service Experience Counters
+* Capella Service Level Metrics
+
+## Capella Service Experience Counters
+
+As initial background to those who are not familiar with the [SDK-RFC on error
+handling](rfc/0058-error-handling.md), Couchbase SDKs attempt to consolidate
+much of the error handling for an app developer into a few specific errors many
+of which may have an error context. The SDKs try to balance this Couchbase
+design of error handling with respecting platform idioms.
+
+That means in SDK instances, we largely have three major errors that a user may
+see along with additional error context.  The error context tends to be service
+specific and these may not have a similar shape as the services were
+independently developed (among other reasons).  There are a number of other
+errors, like InvalidArgument and KeyNotFound which are not useful from a
+Capella Experience perspective.
+
+The three metrics we should collect are:
+
+* Ambiguous Timeout
```

Comment:
What's our unit for 'operation' here? If an operation (as seen from the user's point of view) involves multiple requests sent to the server because of retries or because it's a compound operation, do we count it as one operation?

Reply:
I've just noticed that this is clarified in the example further down:
> 1 is a metric value. Each metric value increment represents a single network operation (e.g. GET_ALL_REPLICAS should be counted as 1 + number_of_replicas).

Reply:
Just a note that it would be a great deal easier from an implementation POV, at least in the JVM SDKs, if we considered an operation to be purely from the user's POV.  E.g. a getAllReplicas as one atomic operation.  We already have such a interception point in the SDKs.

This also makes it simpler to deal with other tricky cases, such as transactions.

Reply:
Compound operations are not visible to the server, and retries as well. So this meter should only track operations visible on the network. I have updated the RFC to point to that.

Reply:
Discussed further in a meeting, since this point significantly complicates the implementation.  Sergey raised the very good point that we want to provide the node, which isn't available at the compound operation level.